#+LATEX_HEADER: \usepackage{grffile}
#+LATEX_HEADER: \usepackage[inline]{enumitem} 
#+LATEX_HEADER: \setdescription{style=multiline,leftmargin=3cm,font=\normalfont}

#+LATEX_HEADER: \usepackage{xcolor}
#+LATEX_HEADER: \hypersetup{
#+LATEX_HEADER:    colorlinks,
#+LATEX_HEADER:    linkcolor={red!50!black},
#+LATEX_HEADER:    citecolor={blue!50!black},
#+LATEX_HEADER:    urlcolor={blue!80!black}
#+LATEX_HEADER:}


#+LATEX_HEADER:\usepackage{setspace}%% The linestretch
#+LATEX_HEADER:\singlespacing

#+LATEX_HEADER:\usepackage[format=hang,indention=0cm,singlelinecheck=true,justification=raggedright,labelfont={normalsize,bf},textfont={normalsize}]{caption} % 


#+LATEX_HEADER:\usepackage{vmargin}
#+LATEX_HEADER:\setpapersize{A4}
#+LATEX_HEADER:\setmarginsrb{2.5cm}{1cm}% links, oben
#+LATEX_HEADER:                                                {2.5cm}{2cm}% rechts, unten
#+LATEX_HEADER:                                                {12pt}{30pt}% Kopf: Höhe, Abstand
#+LATEX_HEADER:                                                {12pt}{30pt}% Fuß: Höhe, AB     
                                                

#+LATEX_HEADER:\usepackage[babel,english=british]{csquotes} 

#+LATEX_HEADER:% English quotes are used.                                       

                                        
#+LATEX_HEADER:\usepackage[english]{babel}                                     

#+LATEX_HEADER:% The diploma thesis will be written in english.


                                
#+LATEX_HEADER:\usepackage[T1]{fontenc}

#+LATEX_HEADER:% Then fontcode is changed to the T1 format.

# #+LATEX_HEADER:\usepackage[latin9]{inputenc}
        

#+LATEX_HEADER: \setlength{\parindent}{0pt}
#+LATEX_HEADER: \setlength{\parskip}{\baselineskip}

# ##+LATEX_HEADER: \usepackage{geometry}
# ##+LATEX_HEADER:  \geometry{
# ##+LATEX_HEADER:  a4paper,
# ##+LATEX_HEADER:  total={210mm,297mm},
# ##+LATEX_HEADER:  left=30mm,
# ##+LATEX_HEADER:  right=30mm,
# ##+LATEX_HEADER:  top=25mm,
# ##+LATEX_HEADER:  bottom=25mm,
# ##+LATEX_HEADER:  }



#+TITLE: *Prerequisites for the bioinformatics-part*
#+AUTHOR: *Alexander Jueterbock, Martin Jakt*
#+DATE: *PhD course: High throughput sequencing of non-model organisms*
#+EMAIL: University of Nordland, Norway
#+OPTIONS: toc:nil H:3 email:t author:t num:nil

# Overview of export options in http://orgmode.org/manual/Export-settings.html#Export-settings


For the practical bioinformatics-part of the HTS course, we will work
remotely on UNIX computers. To connect to these remote computers, you
need to bring your own laptop. *MAC and Linux users* generally don't
need to install any additional software to connect to a remote
computer. However, please install the program [[https://filezilla-project.org][filezilla]]. This program
works on all platforms and helps to transfer files between
computers. 

*MAC* users also need to install [[http://xquartz.macosforge.org/landing/][XQuartz]] for opening applications with
a graphical user interface.

You will have internet access either via eduroam; or, if you bring
your mobile phone, you will get a guest-account password via SMS.

*Windows users*: To connect remotely to a UNIX-based system, download
[[http://www.chiark.greenend.org.uk/~sgtatham/putty/download.html][Putty]] and [[http://sourceforge.net/projects/xming/][Xming]] and follow the configuration guidelines on
http://www.geo.mtu.edu/geoschem/docs/putty_install.html.
Alternatively, you can use [[http://wiki.x2go.org/doku.php/start][X2Go]].


All required programs will be available on our local UNIX computers
and you are not required to install them on your own laptops. However,
if you would like to try them out on your own private computer, here
is a list of programs and scripts that we will use:
* Reading
Please make sure that you have read through the PDF document 'Unix Tools for Bioinformatics' before the 8th of May.
* Programming languages:
- [[http://cran.r-project.org/][R]]
- [[https://www.python.org/][Python]]
- [[https://www.perl.org/][Perl]]

* Trimming and quality control:
- [[http://www.bioinformatics.babraham.ac.uk/projects/fastqc/][Fastqc]]
- [[http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/][TrimGalore!]]
- [[http://hannonlab.cshl.edu/fastx_toolkit/commandline.html][fastx_collapser from FASTX-Toolkit]]
- [[http://sfg.stanford.edu/scripts.html][fastqduplicatecounter.py in the scripts from the Simple Fools Guide]]

* Genome browser:
- [[https://www.broadinstitute.org/igv/][IGV]]

* Genome assembly:
- [[http://mira-assembler.sourceforge.net/][Mira]]

* Mapping and variant calling:
- [[http://bowtie-bio.sourceforge.net/bowtie2/index.shtml][Bowtie2]]
- [[http://marinetics.org/2015/03/03/Bowtie2Filtering.html][Bowtie2Filtering.py]]
- [[http://www.htslib.org/][samtools]]
- [[http://www.htslib.org/][bcftools (with vcfutils.pl)]]
- [[https://broadinstitute.github.io/picard/command-line-overview.html][Picard command line tools]]
- [[http://samtools.sourceforge.net/tabix.shtml][bgzip from tabix]]


